src/arch/src/aarch64/gic/regs/mod.rs

```diff
@@ -0,0 +1,206 @@
+mod dist_regs;
+mod icc_regs;
+mod redist_regs;
+
+use crate::aarch64::gic::{Error, Result};
+use kvm_bindings::kvm_device_attr;
+use kvm_ioctls::DeviceFd;
+use std::iter::StepBy;
+use std::ops::Range;
+use versionize::{VersionMap, Versionize, VersionizeResult};
+use versionize_derive::Versionize;
+
+pub trait MmappedReg {
+    fn range(&self) -> Range<u64>;
+
+    fn iter<T>(&self) -> StepBy<Range<u64>>
+    where
+        Self: Sized,
+    {
+        self.range().step_by(std::mem::size_of::<T>())
+    }
+}
+
+pub trait VgicRegEngine {
+    type Reg: MmappedReg;
+    type RegChunk: Default + Clone;
+
+    fn group() -> u32;
+
+    fn mpidr_mask() -> u64;
+
+    fn kvm_device_attr(offset: u64, val: &mut Self::RegChunk, mpidr: u64) -> kvm_device_attr {
+        kvm_device_attr {
+            group: Self::group(),
+            attr: (mpidr & Self::mpidr_mask()) | offset,
+            addr: val as *mut Self::RegChunk as u64,
+            flags: 0,
+        }
+    }
+
+    #[inline]
+    fn get_reg_data(fd: &DeviceFd, reg: &Self::Reg, mpidr: u64) -> Result<Vec<Self::RegChunk>>
+    where
+        Self: Sized,
+    {
+        let mut data = Vec::with_capacity(reg.iter::<Self::RegChunk>().count());
+        for offset in reg.iter::<Self::RegChunk>() {
```

Comment:
nit:

don't build iter twice - although most likely the compiler already does this optimization for us anyway

```suggestion
        let iter = reg.iter::<Self::RegChunk>();
        let mut data = Vec::with_capacity(iter.count());
        for offset in iter {
```